trade-finance-manager-ui/server/controllers/login/loginService.js

```diff
@@ -0,0 +1,67 @@
+const { SSO } = require("../../constants");
+
+/**
+ * verifyReferrerForExternalSsoPost
+ * Verify referrer, allow empty referrer for localhost.
+ * 1) If host is localhost then referrer will be missing, no need to check it
+ * 2) Return error message if referrer is missing or it is not from SSO authority
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
+ */
+const verifyReferrerForExternalSsoPost = (req) => {
+  const hostnameWithoutPort = req.get('host').split(':')[0];
+  const referrer = req.get('Referrer');
+
+  // Referrer is not available in localhost because of policy "no-referrer-when-downgrade".
+  if (hostnameWithoutPort === 'localhost') {
+    return;
+  }
+
+  if (referrer && referrer.indexOf(`${SSO.AUTHORITY}/`) === 0) {
+    return;
+  }
+
+  console.error('Login request comming from unexpected website: %s', referrer);
+  throw new Error('Login request comming from unexpected website.');
+}
+
+/**
+ * verifyBodyForExternalSsoPost
+ * Verify SSO input parameters, allow limited/safe number of characters.
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
+ */
+const verifyBodyForExternalSsoPost = (templateParams) => {
+  const paramsWithNotAllowedCharacters = Object.entries(templateParams).filter(([, value]) => (/^[0-9a-zA-Z-_.]*$/.test(value) === false));
+
+  if (paramsWithNotAllowedCharacters.length) {
+    console.error('Login request data contains unexpected characters in: %O', paramsWithNotAllowedCharacters);
+    throw new Error('Login request data contains unexpected characters.');
+  }
+}
+
+/**
+ * acceptExternalSsoPost
+ * Accept SSO POST from SSO Authority, build form that will auto submit to have session cookie.
+ * 1) Verify referrer.
+ * 2) Verify Post body to have safe parameters.
+ * 3) Return page with form acceptExternalSsoPostForm that will autosubmit.
+ * @param {Object} req: Express request
+ * @param {Object} res: Express response
+ * @returns {Object} response with rendered sso/accept-external-sso-post.njk.
+ */
+const acceptExternalSsoPost = (req, res) => {
+  const { code, client_info: clientInfo, state, session_state: sessionState } = req.body;
+  const externalTemplateParams = { code, clientInfo, state, sessionState };
+  try {
+    verifyReferrerForExternalSsoPost(req);
+    verifyBodyForExternalSsoPost(externalTemplateParams);
+
+    return res.render('sso/accept-external-sso-post.njk', { ...externalTemplateParams, azureSsoAuthority: `${SSO.AUTHORITY}/`});
+  } catch(error) {
+    const errorMessage = error.message || `Login process failed - try again or contact us using details bellow.`;
```

Comment:
same comment about logging errors from `index.js` here and returning a standardised error msg :)

Reply:
I added logging. 
"Verify" functions return quite user friendly messages, I am tempted to leave output as it is.